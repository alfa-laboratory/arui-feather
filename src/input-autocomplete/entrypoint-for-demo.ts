// REMOVE AFTER UPGRADE STYLEGUIDIST
export { default } from './index';
