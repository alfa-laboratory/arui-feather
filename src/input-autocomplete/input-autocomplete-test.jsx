/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { render, cleanUp, simulate } from '../test-utils';

import InputAutocomplete from './input-autocomplete';

import { SCROLL_TO_CORRECTION } from '../vars';

const OPTIONS = [
    { value: 'Vkontakte' },
    { value: 'Facebook' },
    { value: 'Twitter' }
];

function renderInputAutocomplete(props = {}) {
    let inputAutocomplete = render(<InputAutocomplete { ...props } />);

    let inputNode = inputAutocomplete.node.querySelector('.input');
    let controlNode = inputAutocomplete.node.querySelector('input');
    let popupNode = document.querySelector('.popup');

    return { inputAutocomplete, inputNode, controlNode, popupNode };
}

describe('input-autocomplete', () => {
    let originalWindowScrollTo = window.scrollTo;

    beforeEach(() => {
        window.scrollTo = chai.spy();
    });

    afterEach(() => {
        cleanUp();
        window.scrollTo = originalWindowScrollTo;
    });

    it('should render without problem', () => {
        let { inputAutocomplete } = renderInputAutocomplete();

        expect(inputAutocomplete.node).to.exist;
    });

    it('should render input and popup with options', () => {
        let { popupNode, controlNode } = renderInputAutocomplete({ options: OPTIONS });

        expect(popupNode).to.have.class('popup');
        expect(controlNode).to.have.class('input__control');
    });

    it('should set class on public focus method', (done) => {
        let { inputAutocomplete, inputNode } = renderInputAutocomplete();

        inputAutocomplete.instance.focus();

        setTimeout(() => {
            expect(inputNode).to.have.class('input_focused');
            done();
        }, 0);
    });

    it('should scroll window to element on public scrollTo method', (done) => {
        let { inputAutocomplete } = renderInputAutocomplete();
        let elemTopPosition = inputAutocomplete.node.getBoundingClientRect().top;
        let elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        inputAutocomplete.instance.scrollTo();

        setTimeout(() => {
            expect(window.scrollTo).to.have.been.called.with(0, elemScrollTo);
            done();
        }, 0);
    });

    it('should unset class on input blur', (done) => {
        let { inputAutocomplete, inputNode } = renderInputAutocomplete();

        inputAutocomplete.instance.focus();

        expect(inputNode).to.have.class('input_focused');

        setTimeout(() => {
            inputAutocomplete.instance.blur();

            setTimeout(() => {
                expect(inputNode).to.not.have.class('input_focused');
                done();
            }, 0);
        }, 0);
    });

    it('should set width to popup equal or more than button width', () => {
        let { popupNode, inputNode } = renderInputAutocomplete({ options: OPTIONS, opened: true });

        let inputWidth = inputNode.getBoundingClientRect().width;
        let popupWidth = popupNode.getBoundingClientRect().width;

        expect(popupWidth).to.be.at.least(inputWidth);
    });

    it('should render all options when input value is empty', () => {
        let { popupNode } = renderInputAutocomplete({ options: OPTIONS });
        let optionsNode = popupNode.querySelectorAll('.menu-item');

        expect(optionsNode.length).is.equal(OPTIONS.length);
    });

    it('should change input value after option was clicked', (done) => {
        let { popupNode, controlNode } = renderInputAutocomplete({ options: OPTIONS });
        let firstOptionNode = popupNode.querySelector('.menu-item');

        firstOptionNode.click();

        setTimeout(() => {
            expect(controlNode).to.have.attr('value', OPTIONS[0].value);
            done();
        }, 0);
    });

    it('should call `onItemSelect` callback after option was clicked', () => {
        let onItemSelect = chai.spy();
        let { popupNode } = renderInputAutocomplete({ onItemSelect, options: OPTIONS });
        let firstOptionNode = popupNode.querySelector('.menu-item');

        firstOptionNode.click();

        expect(onItemSelect).to.have.been.called.once;
    });

    it('should call `onChange` callback after option was clicked and pass value to it', () => {
        let onChange = chai.spy();
        let { popupNode } = renderInputAutocomplete({ onChange, options: OPTIONS });
        let firstOptionNode = popupNode.querySelector('.menu-item');

        firstOptionNode.click();

        expect(onChange).to.have.been.called.with('Vkontakte');
    });

    it('should call `onKeyDown` callback after key down in input', () => {
        let onKeyDown = chai.spy();
        let { controlNode } = renderInputAutocomplete({ onKeyDown });

        simulate(controlNode, 'keyDown');

        expect(onKeyDown).to.have.been.called.once;
    });

    it('should call `onFocus` callback after component was focused', (done) => {
        let onFocus = chai.spy();
        let { inputAutocomplete } = renderInputAutocomplete({ onFocus });

        inputAutocomplete.instance.focus();

        setTimeout(() => {
            expect(onFocus).to.have.been.called.once;
            done();
        }, 0);
    });

    it('should call `onBlur` callback after component was blured', (done) => {
        let onBlur = chai.spy();
        let { inputAutocomplete } = renderInputAutocomplete({ onBlur });

        inputAutocomplete.instance.focus();

        setTimeout(() => {
            inputAutocomplete.instance.blur();

            setTimeout(() => {
                expect(onBlur).to.have.been.called.once;
                done();
            }, 0);
        }, 0);
    });

    it('should call `onChange` callback', () => {
        let onChange = chai.spy();
        let { controlNode } = renderInputAutocomplete({ onChange });

        simulate(controlNode, 'change', { target: { value: 'other value' } });

        expect(onChange).to.have.been.called.once;
    });
});
