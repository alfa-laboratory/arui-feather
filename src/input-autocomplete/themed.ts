import { withTheme } from '../cn';
import Component from './input-autocomplete';

export default withTheme(Component);
