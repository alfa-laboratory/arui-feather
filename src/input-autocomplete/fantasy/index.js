import '../../input/fantasy';
import '../../menu/fantasy';
import '../../popup/fantasy';
import '../../resize-sensor';

import './input-autocomplete.css';

export default from '../input-autocomplete';
