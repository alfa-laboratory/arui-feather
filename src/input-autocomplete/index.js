import '../input';
import '../menu';
import '../popup';
import '../resize-sensor';

import './input-autocomplete.css';

export default from './input-autocomplete';
