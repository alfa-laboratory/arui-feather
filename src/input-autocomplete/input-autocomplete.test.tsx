/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
/* eslint import/no-extraneous-dependencies: [2, {"devDependencies": true}] */

import React from 'react';
import { mount } from 'enzyme';

import { Input } from '../input/input';
import { InputAutocomplete } from './input-autocomplete';

import { SCROLL_TO_CORRECTION } from '../vars';

const OPTIONS = [
    { value: 'VKontakte' },
    { value: 'Facebook' },
    { value: 'Twitter' },
];

const OPTIONS2 = [
    { value: 'VKontakte' },
    {
        value: 'VKontakte',
        key: 'vk2',
    },
];

describe('input-autocomplete', () => {
    const originalWindowScrollTo = window.scrollTo;

    beforeEach(() => {
        window.scrollTo = jest.fn();
    });

    afterEach(() => {
        window.scrollTo = originalWindowScrollTo;
    });

    it('should render without problem', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete />);

        expect(inputAutocomplete).toMatchSnapshot();
    });

    it('should render without problem when give item with duplicate value', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS2 } />);
        const optionsNode = inputAutocomplete.find('.menu-item');

        expect(optionsNode.length).toBe(OPTIONS2.length);
    });

    it('should render input with provided default value', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } defaultValue="default" />);

        expect(inputAutocomplete).toMatchSnapshot();
    });

    it('should render input and popup with options', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } />);

        expect(inputAutocomplete.find('.popup').length).toBe(1);
        expect(inputAutocomplete.find('.input__control').length).toBe(1);
    });

    it('should call input.focus on public focus method', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } />);
        const inputInstance = inputAutocomplete.find<Input>(Input).instance();

        jest.spyOn(inputInstance, 'focus');

        inputAutocomplete.instance().focus();

        expect(inputInstance.focus).toHaveBeenCalled();
    });

    it('should scroll window to element on public scrollTo method', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete />);
        const inputNode = inputAutocomplete.find('.input__control');

        const elemTopPosition = inputNode.getDOMNode().getBoundingClientRect().top;
        const elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        inputAutocomplete.instance().scrollTo();

        expect(window.scrollTo).toHaveBeenCalledWith(0, elemScrollTo);
    });

    it('should unset class on input blur', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete />);

        inputAutocomplete.find('input').simulate('focus');

        expect(inputAutocomplete.find('.input_focused').length).toBe(1);

        inputAutocomplete.find('input').simulate('blur');

        expect(inputAutocomplete.find('.input_focused').length).toBe(0);
    });

    it('should set width to popup equal or more than button width', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } opened={ true } />);
        const inputNode = inputAutocomplete.find('.input').at(0);
        const popupNode = inputAutocomplete.find('.popup');

        const inputWidth = inputNode.getDOMNode().getBoundingClientRect().width;
        const popupWidth = popupNode.getDOMNode().getBoundingClientRect().width;

        expect(popupWidth).toBeGreaterThanOrEqual(inputWidth);
    });

    it('should set popup width equal to input width when equalPopupWidth = true', () => {
        const props = {
            options: [
                {
                    value: `Very, very long option text
                            used just to make autocomplete popup
                            strech really really wide and another
                            couple of words just to be sure`,
                },
            ],
            equalPopupWidth: true,
            opened: true,
        };

        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete { ...props } />);
        const inputNode = inputAutocomplete.find('.input').at(0);
        const popupNode = inputAutocomplete.find('.popup');
        const popupWidth = popupNode.getDOMNode().getBoundingClientRect().width;
        const inputWidth = inputNode.getDOMNode().getBoundingClientRect().width;

        expect(popupWidth).toBe(inputWidth);
    });

    it('should set directions to popup', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } directions={ ['right-bottom'] } />);
        const popupNode = inputAutocomplete.find('.popup');

        expect(popupNode.props().className).toContain('popup_direction_right-bottom');
    });

    it('should render all options when input value is empty', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } />);
        const optionsNode = inputAutocomplete.find('.menu-item');

        expect(optionsNode.length).toBe(OPTIONS.length);
    });

    it('should change input value after option was clicked', () => {
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete options={ OPTIONS } />);
        const firstOptionNode = inputAutocomplete.find('.menu-item').at(0);

        firstOptionNode.simulate('click');

        expect(inputAutocomplete.find('input').props().value).toBe(OPTIONS[0].value);
    });

    it('should call `onItemSelect` callback after option was clicked', () => {
        const onItemSelect = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(
            <InputAutocomplete options={ OPTIONS } onItemSelect={ onItemSelect } />,
        );
        const firstOptionNode = inputAutocomplete.find('.menu-item').at(0);

        firstOptionNode.simulate('click');

        expect(onItemSelect).toHaveBeenCalled();
    });

    it('should call `onChange` callback after option was clicked and pass value to it', () => {
        const onChange = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(
            <InputAutocomplete options={ OPTIONS } onChange={ onChange } />,
        );
        const firstOptionNode = inputAutocomplete.find('.menu-item').at(0);

        firstOptionNode.simulate('click');

        expect(onChange).toHaveBeenCalledWith(
            'VKontakte',
            expect.objectContaining({
                nativeEvent: expect.any(Object),
            }),
        );
    });

    it('should call `onChange` callback after option was clicked and pass `text` when `option.text` is exist', () => {
        const onChange = jest.fn();
        const OPTIONS = [
            { value: 'VKontakte', text: 'ВКонтакте' },
            { value: 'Facebook', text: 'Фейсбук' },
            { value: 'Twitter', text: 'Твиттер' },
        ];
        const inputAutocomplete = mount<InputAutocomplete>(
            <InputAutocomplete options={ OPTIONS } onChange={ onChange } />,
        );

        const firstOptionNode = inputAutocomplete.find('.menu-item').at(0);

        firstOptionNode.simulate('click');

        expect(onChange).toHaveBeenCalledWith(
            'ВКонтакте',
            expect.objectContaining({
                nativeEvent: expect.any(Object),
            }),
        );
    });

    it('should not call `onChange` callback after option was clicked when updateValueOnItemSelect = false', () => {
        const onChange = jest.fn();
        const updateValueOnItemSelect = mount(
            <InputAutocomplete options={ OPTIONS } onChange={ onChange } updateValueOnItemSelect={ false } />,
        );
        const firstOptionNode = updateValueOnItemSelect.find('.menu-item').at(0);

        firstOptionNode.simulate('click');

        expect(onChange).not.toHaveBeenCalled();
    });

    it('should call `onKeyDown` callback after key down in input', () => {
        const onKeyDown = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete onKeyDown={ onKeyDown } />);
        const controlNode = inputAutocomplete.find('input');

        controlNode.simulate('keyDown');

        expect(onKeyDown).toHaveBeenCalled();
    });

    it('should call `onFocus` callback after component was focused', () => {
        const onFocus = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete onFocus={ onFocus } />);
        const controlNode = inputAutocomplete.find('input');

        // simulated event don't actually change focus
        // so we just patch input.getControl
        ((inputAutocomplete.instance() as any).input).getControl = jest.fn().mockReturnValue(document.activeElement);

        controlNode.simulate('focus');

        expect(onFocus).toHaveBeenCalled();
    });

    it('should call `onBlur` callback after component was blured', (done) => {
        const onBlur = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete onBlur={ onBlur } />);
        const controlNode = inputAutocomplete.find('input');

        inputAutocomplete.setState({ inputFocused: true });
        controlNode.simulate('blur');

        setTimeout(() => {
            expect(onBlur).toHaveBeenCalled();
            done();
        }, 0);
    });

    it('should call `onChange` callback', () => {
        const onChange = jest.fn();
        const inputAutocomplete = mount<InputAutocomplete>(<InputAutocomplete onChange={ onChange } />);
        const controlNode = inputAutocomplete.find('input');

        controlNode.simulate('change', { target: { value: 'other value' } });

        expect(onChange).toHaveBeenCalled();
    });

    it('should close popup after item select if closeOnSelect is set', (done) => {
        const inputAutocomplete = mount<InputAutocomplete>(
            <InputAutocomplete closeOnSelect={ true } updateValueOnItemSelect={ false } options={ OPTIONS } />,
        );
        const inputNode = (inputAutocomplete.instance() as any).input;

        jest.spyOn(inputNode, 'blur');

        inputAutocomplete.setState({ inputFocused: true });

        expect(inputAutocomplete.find('.popup').props().className).toContain('popup_visible');

        inputAutocomplete.find('.menu-item').at(0).simulate('click');

        setTimeout(() => {
            expect(inputNode.blur).toHaveBeenCalled();
            done();
        }, 0);
    });

    it('should have disabled option, which can not be selected, if disabled property is set true', (done) => {
        const OPTIONS = [
            { value: 'VKontakte', props: { disabled: true } },
            { value: 'Facebook' },
            { value: 'Twitter' },
        ];
        const onChange = jest.fn();

        const inputAutocomplete = mount(
            <InputAutocomplete options={ OPTIONS } onChange={ onChange } />,
        );

        inputAutocomplete.find('.menu-item').at(0).simulate('click');

        const firstOptionNode = inputAutocomplete.find('.menu-item').at(0);

        expect(firstOptionNode.props().className).toContain('menu-item_disabled');

        setTimeout(() => {
            expect(onChange).toHaveBeenCalledTimes(0);
            done();
        }, 0);
    });
});
