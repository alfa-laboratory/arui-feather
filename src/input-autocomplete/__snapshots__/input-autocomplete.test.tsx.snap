// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input-autocomplete should render input with provided default value 1`] = `
<InputAutocomplete
  closeOnSelect={false}
  defaultValue="default"
  directions={
    Array [
      "bottom-left",
      "bottom-right",
      "top-left",
      "top-right",
    ]
  }
  disabled={false}
  equalPopupWidth={false}
  options={
    Array [
      Object {
        "value": "VKontakte",
      },
      Object {
        "value": "Facebook",
      },
      Object {
        "value": "Twitter",
      },
    ]
  }
  renderPopupOnFocus={false}
  size="m"
  updateValueOnItemSelect={true}
  width="default"
>
  <div
    className="input__autocomplete-case input__autocomplete-case_width_default"
  >
    <ForwardRef(ThemedComponent(Input))
      autocomplete={false}
      className="input input_has-autocomplete"
      closeOnSelect={false}
      defaultValue="default"
      directions={
        Array [
          "bottom-left",
          "bottom-right",
          "top-left",
          "top-right",
        ]
      }
      disabled={false}
      equalPopupWidth={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      options={
        Array [
          Object {
            "value": "VKontakte",
          },
          Object {
            "value": "Facebook",
          },
          Object {
            "value": "Twitter",
          },
        ]
      }
      renderPopupOnFocus={false}
      size="m"
      updateValueOnItemSelect={true}
      value="default"
      width="default"
    >
      <Input
        autocomplete={false}
        className="input input_has-autocomplete"
        closeOnSelect={false}
        defaultValue="default"
        directions={
          Array [
            "bottom-left",
            "bottom-right",
            "top-left",
            "top-right",
          ]
        }
        disabled={false}
        equalPopupWidth={false}
        formNoValidate={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        options={
          Array [
            Object {
              "value": "VKontakte",
            },
            Object {
              "value": "Facebook",
            },
            Object {
              "value": "Twitter",
            },
          ]
        }
        renderPopupOnFocus={false}
        resetError={true}
        size="m"
        theme="alfa-on-white"
        type="text"
        updateValueOnItemSelect={true}
        value="default"
        view="default"
        width="default"
      >
        <span
          className="input input input_has-autocomplete input_type_text input_view_default input_size_m input_width_default input_has-value input_theme_alfa-on-white"
        >
          <span
            className="input__inner"
          >
            <span
              className="input__box"
              key="input-wrapper"
            >
              <span
                className="input__input-wrapper"
              >
                <input
                  autoComplete="off"
                  className="input__control"
                  formNoValidate={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onPaste={[Function]}
                  onTouchCancel={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  type="text"
                  value="default"
                  view="default"
                />
              </span>
            </span>
          </span>
        </span>
      </Input>
    </ForwardRef(ThemedComponent(Input))>
    <ResizeSensor
      key="popup-sensor"
      onResize={[Function]}
    >
      <iframe
        style={
          Object {
            "background": "transparent",
            "border": "none",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
            "zIndex": -1,
          }
        }
        tabIndex={-1}
      />
    </ResizeSensor>
    <ForwardRef(ThemedComponent(Popup))
      className="input__popup"
      directions={
        Array [
          "bottom-left",
          "bottom-right",
          "top-left",
          "top-right",
        ]
      }
      height="adaptive"
      key="popup"
      minWidth={0}
      onClickOutside={[Function]}
      padded={false}
      size="m"
      target="anchor"
      visible={false}
    >
      <Popup
        className="input__popup"
        directions={
          Array [
            "bottom-left",
            "bottom-right",
            "top-left",
            "top-right",
          ]
        }
        fitContaiterOffset={0}
        height="adaptive"
        minWidth={0}
        onClickOutside={[Function]}
        padded={false}
        secondaryOffset={0}
        size="m"
        target="anchor"
        theme="alfa-on-white"
        visible={false}
      >
        <Portal
          containerInfo={
            <body>
              <div
                class="popup input__popup popup_direction_bottom-left popup_target_anchor popup_size_m popup_height_adaptive popup_theme_alfa-on-white"
                style="top: 0px; left: 0px; height: auto; min-width: 0; max-width: none; max-height: none;"
              >
                <div
                  class="popup__container"
                >
                  <div
                    class="popup__inner"
                  >
                    <div
                      class="popup__content"
                    >
                      <div
                        class="menu input__menu menu_size_m menu_group-view_default menu_mode_radio-check menu_theme_alfa-on-white"
                        tabindex="0"
                      >
                        <div
                          class="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                        >
                          <span
                            class="menu-item__control"
                          >
                            VKontakte
                          </span>
                        </div>
                        <div
                          class="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                        >
                          <span
                            class="menu-item__control"
                          >
                            VKontakte
                          </span>
                        </div>
                      </div>
                      <iframe
                        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; background: transparent; z-index: -1;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="popup input__popup popup_direction_bottom-left popup_target_anchor popup_size_m popup_height_adaptive popup_theme_alfa-on-white"
                style="top: 0px; left: 0px; height: auto; min-width: 0; max-width: none; max-height: none;"
              >
                <div
                  class="popup__container"
                >
                  <div
                    class="popup__inner"
                  >
                    <div
                      class="popup__content"
                    >
                      <div
                        class="menu input__menu menu_size_m menu_group-view_default menu_mode_radio-check menu_theme_alfa-on-white"
                        tabindex="0"
                      >
                        <div
                          class="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                        >
                          <span
                            class="menu-item__control"
                          >
                            VKontakte
                          </span>
                        </div>
                        <div
                          class="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                        >
                          <span
                            class="menu-item__control"
                          >
                            Facebook
                          </span>
                        </div>
                        <div
                          class="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                        >
                          <span
                            class="menu-item__control"
                          >
                            Twitter
                          </span>
                        </div>
                      </div>
                      <iframe
                        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; background: transparent; z-index: -1;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </body>
          }
        >
          <div
            className="popup input__popup popup_direction_bottom-left popup_target_anchor popup_size_m popup_height_adaptive popup_theme_alfa-on-white"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "bottom": undefined,
                "height": "auto",
                "left": 0,
                "maxHeight": "none",
                "maxWidth": "none",
                "minWidth": 0,
                "right": undefined,
                "top": 0,
              }
            }
          >
            <div
              className="popup__container"
            >
              <div
                className="popup__inner"
                onScroll={[Function]}
              >
                <div
                  className="popup__content"
                >
                  <ForwardRef(ThemedComponent(Menu))
                    autoFocusFirstItem={true}
                    checkedItems={Array []}
                    className="input__menu"
                    content={
                      Array [
                        Object {
                          "content": "VKontakte",
                          "key": "VKontakte",
                          "props": undefined,
                          "value": "VKontakte",
                        },
                        Object {
                          "content": "Facebook",
                          "key": "Facebook",
                          "props": undefined,
                          "value": "Facebook",
                        },
                        Object {
                          "content": "Twitter",
                          "key": "Twitter",
                          "props": undefined,
                          "value": "Twitter",
                        },
                      ]
                    }
                    highlightedItem={null}
                    mode="radio-check"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onHighlightItem={[Function]}
                    onItemCheck={[Function]}
                    onKeyDown={[Function]}
                    size="m"
                  >
                    <Menu
                      autoFocusFirstItem={true}
                      checkedItems={Array []}
                      className="input__menu"
                      content={
                        Array [
                          Object {
                            "content": "VKontakte",
                            "key": "VKontakte",
                            "props": undefined,
                            "value": "VKontakte",
                          },
                          Object {
                            "content": "Facebook",
                            "key": "Facebook",
                            "props": undefined,
                            "value": "Facebook",
                          },
                          Object {
                            "content": "Twitter",
                            "key": "Twitter",
                            "props": undefined,
                            "value": "Twitter",
                          },
                        ]
                      }
                      groupView="default"
                      highlightedItem={null}
                      mode="radio-check"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onHighlightItem={[Function]}
                      onItemCheck={[Function]}
                      onKeyDown={[Function]}
                      size="m"
                      theme="alfa-on-white"
                    >
                      <div
                        className="menu input__menu menu_size_m menu_group-view_default menu_mode_radio-check menu_theme_alfa-on-white"
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        tabIndex={0}
                      >
                        <ForwardRef(ThemedComponent(MenuItem))
                          checked={false}
                          hovered={null}
                          key="VKontakte"
                          onClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          size="m"
                          type="block"
                          value="VKontakte"
                        >
                          <MenuItem
                            checked={false}
                            hovered={null}
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            onMouseLeave={[Function]}
                            size="m"
                            theme="alfa-on-white"
                            type="block"
                            value="VKontakte"
                            view="plain"
                          >
                            <div
                              className="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                            >
                              <span
                                className="menu-item__control"
                              >
                                VKontakte
                              </span>
                            </div>
                          </MenuItem>
                        </ForwardRef(ThemedComponent(MenuItem))>
                        <ForwardRef(ThemedComponent(MenuItem))
                          checked={false}
                          hovered={null}
                          key="Facebook"
                          onClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          size="m"
                          type="block"
                          value="Facebook"
                        >
                          <MenuItem
                            checked={false}
                            hovered={null}
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            onMouseLeave={[Function]}
                            size="m"
                            theme="alfa-on-white"
                            type="block"
                            value="Facebook"
                            view="plain"
                          >
                            <div
                              className="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                            >
                              <span
                                className="menu-item__control"
                              >
                                Facebook
                              </span>
                            </div>
                          </MenuItem>
                        </ForwardRef(ThemedComponent(MenuItem))>
                        <ForwardRef(ThemedComponent(MenuItem))
                          checked={false}
                          hovered={null}
                          key="Twitter"
                          onClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          size="m"
                          type="block"
                          value="Twitter"
                        >
                          <MenuItem
                            checked={false}
                            hovered={null}
                            onClick={[Function]}
                            onMouseEnter={[Function]}
                            onMouseLeave={[Function]}
                            size="m"
                            theme="alfa-on-white"
                            type="block"
                            value="Twitter"
                            view="plain"
                          >
                            <div
                              className="menu-item menu-item_type_block menu-item_theme_alfa-on-white"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                            >
                              <span
                                className="menu-item__control"
                              >
                                Twitter
                              </span>
                            </div>
                          </MenuItem>
                        </ForwardRef(ThemedComponent(MenuItem))>
                      </div>
                    </Menu>
                  </ForwardRef(ThemedComponent(Menu))>
                  <ResizeSensor
                    onResize={[Function]}
                  >
                    <iframe
                      style={
                        Object {
                          "background": "transparent",
                          "border": "none",
                          "height": "100%",
                          "left": 0,
                          "position": "absolute",
                          "top": 0,
                          "width": "100%",
                          "zIndex": -1,
                        }
                      }
                      tabIndex={-1}
                    />
                  </ResizeSensor>
                </div>
              </div>
            </div>
          </div>
        </Portal>
      </Popup>
    </ForwardRef(ThemedComponent(Popup))>
  </div>
</InputAutocomplete>
`;

exports[`input-autocomplete should render without problem 1`] = `
<InputAutocomplete
  closeOnSelect={false}
  directions={
    Array [
      "bottom-left",
      "bottom-right",
      "top-left",
      "top-right",
    ]
  }
  disabled={false}
  equalPopupWidth={false}
  options={Array []}
  renderPopupOnFocus={false}
  size="m"
  updateValueOnItemSelect={true}
  width="default"
>
  <div
    className="input__autocomplete-case input__autocomplete-case_width_default"
  >
    <ForwardRef(ThemedComponent(Input))
      autocomplete={false}
      className="input input_has-autocomplete"
      closeOnSelect={false}
      directions={
        Array [
          "bottom-left",
          "bottom-right",
          "top-left",
          "top-right",
        ]
      }
      disabled={false}
      equalPopupWidth={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      options={Array []}
      renderPopupOnFocus={false}
      size="m"
      updateValueOnItemSelect={true}
      value=""
      width="default"
    >
      <Input
        autocomplete={false}
        className="input input_has-autocomplete"
        closeOnSelect={false}
        directions={
          Array [
            "bottom-left",
            "bottom-right",
            "top-left",
            "top-right",
          ]
        }
        disabled={false}
        equalPopupWidth={false}
        formNoValidate={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        options={Array []}
        renderPopupOnFocus={false}
        resetError={true}
        size="m"
        theme="alfa-on-white"
        type="text"
        updateValueOnItemSelect={true}
        value=""
        view="default"
        width="default"
      >
        <span
          className="input input input_has-autocomplete input_type_text input_view_default input_size_m input_width_default input_theme_alfa-on-white"
        >
          <span
            className="input__inner"
          >
            <span
              className="input__box"
              key="input-wrapper"
            >
              <span
                className="input__input-wrapper"
              >
                <input
                  autoComplete="off"
                  className="input__control"
                  formNoValidate={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onPaste={[Function]}
                  onTouchCancel={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  type="text"
                  value=""
                  view="default"
                />
              </span>
            </span>
          </span>
        </span>
      </Input>
    </ForwardRef(ThemedComponent(Input))>
  </div>
</InputAutocomplete>
`;
